File: raffia-core/src/main/java/com/codepoetics/raffia/paths/PathSegment.java
package com.codepoetics.raffia.paths;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.baskets.Visitor;

import java.util.List;

public interface PathSegment {

  Visitor<Basket> createUpdater(Path path, Visitor<Basket> updater);
  Visitor<Basket> createItemUpdater(Path tail, Visitor<Basket> updater);

  <T> Visitor<List<T>> createProjector(Path tail, Visitor<List<T>> projector);
  Visitor<List<Basket>> createItemProjector(Path tail);

  PathSegmentMatchResult matchesIndex(int index);

  PathSegmentMatchResult matchesKey(String key);

  boolean isConditional();

  String representation();
}

File: raffia-core/src/main/java/com/codepoetics/raffia/paths/PathSegment.kt
package com.codepoetics.raffia.paths

import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.java.api.Projector
import com.codepoetics.raffia.java.api.Updater

interface PathSegment {

    fun createUpdater(tail: Path, updater: Updater): Updater
    fun createItemUpdater(tail: Path, updater: Updater): Updater

    fun createProjector(tail: Path): Projector<Basket>
    fun createItemProjector(tail: Path): Projector<Basket>

    fun matchesIndex(index: Int): PathSegmentMatchResult

    fun matchesKey(key: String): PathSegmentMatchResult

    val isConditional: Boolean

    fun representation(): String
}
